// @bunli/runtime/src/components/index.ts
export {
  ThemeProvider,
  createTheme,
  useTuiTheme,
  darkThemeTokens,
  lightThemeTokens,
  type ThemeProviderProps,
  type TuiTheme,
  type TuiThemeInput,
  type TuiThemeTokens,
} from "./theme.js";

export {
  createKeyMatcher,
  matchesKeyBinding,
  eventToBinding,
  type KeyBinding,
  type KeymapDefinition,
  type KeyMatcher,
} from "./keymap.js";

export {
  displayWidth,
  truncateEnd,
  padEndTo,
  formatFixedWidth,
  type TextOverflowMode,
} from "./text-layout.js";

export {
  FocusScopeProvider,
  useFocusScope,
  useScopedKeyboard,
  dispatchScopedKeyboardEvent,
  type FocusScopeProviderProps,
  type ScopedKeyHandler,
  type UseFocusScopeOptions,
  type UseScopedKeyboardOptions,
} from "./focus-scope.js";

export {
  OverlayHostProvider,
  OverlayPortal,
  type OverlayHostProviderProps,
  type OverlayPortalProps,
} from "./overlay-host.js";

export {
  DialogProvider,
  useDialogManager,
  DialogDismissedError,
  sortDialogs,
  getTopDialog,
  getSelectableIndices,
  getResolvedChooseIndex,
  getAdjacentSelectableIndex,
  type ChooseDialogOption,
  type ChooseDialogOptions,
  type ConfirmDialogOptions,
  type DialogManager,
  type DialogOpenOptions,
  type DialogProviderProps,
  type DialogRenderContext,
} from "./dialog-manager.js";
