// @bunli/tui/src/components/spinner.tsx
import { useEffect, useState } from 'react'
import { useTuiTheme } from '@bunli/runtime/app'

export type SpinnerVariant = 'line' | 'dot' | 'minidot' | 'jump' | 'pulse' | 'points' | 'globe' | 'moon' | 'monkey' | 'meter' | 'hamburger'

export const SPINNERS: Record<SpinnerVariant, { frames: string[]; interval: number }> = {
  line:      { frames: ['|', '/', '-', '\\'], interval: 130 },
  dot:       { frames: ['в Ӣ', 'в ҷ', 'в №', 'в ё', 'в ј', 'в ҙ', 'в Ұ', 'в §', 'в Ү', 'в Ҹ'], interval: 80 },
  minidot:   { frames: ['в „', 'в Ӯ', 'в Ғ', 'в Ҳ', 'в җ', 'в  '], interval: 100 },
  jump:      { frames: ['вўҖв Җ', 'вЎҖв Җ', 'в „в Җ', 'вўӮв Җ', 'вЎӮв Җ', 'в …в Җ', 'вўғв Җ', 'вЎғв Җ', 'в Қв Җ', 'вўӢв Җ', 'вЎӢв Җ', 'в Қв Ғ', 'вўӢв Ғ', 'вЎӢв Ғ', 'в Қв ү', 'в Ӣв ү', 'в Ӣв ү', 'в үв ҷ', 'в үв ҷ', 'в үв ©', 'в Ҳвўҷ', 'в ҲвЎҷ', 'вўҲв ©', 'вЎӮв ©', 'в …в ©', 'вўғв ©', 'вЎғв ©', 'в Қв ©', 'вўӢв ©', 'вЎӢв ©', 'в Қв ©', 'вўӢв ©', 'вЎӢв ©', 'в Қвўү', 'в ҚвЎү', 'в Қв Ӣ'], interval: 100 },
  pulse:     { frames: ['в–Ҳ', 'в–“', 'в–’', 'в–‘', 'в–’', 'в–“'], interval: 120 },
  points:    { frames: ['вҲҷвҲҷвҲҷ', 'в—ҸвҲҷвҲҷ', 'вҲҷв—ҸвҲҷ', 'вҲҷвҲҷв—Ҹ'], interval: 200 },
  globe:     { frames: ['рҹҢҚ', 'рҹҢҺ', 'рҹҢҸ'], interval: 200 },
  moon:      { frames: ['рҹҢ‘', 'рҹҢ’', 'рҹҢ“', 'рҹҢ”', 'рҹҢ•', 'рҹҢ–', 'рҹҢ—', 'рҹҢҳ'], interval: 120 },
  monkey:    { frames: ['рҹҷҲ', 'рҹҷү', 'рҹҷҠ'], interval: 300 },
  meter:     { frames: ['в–ұв–ұв–ұв–ұв–ұв–ұв–ұ', 'в–°в–ұв–ұв–ұв–ұв–ұв–ұ', 'в–°в–°в–ұв–ұв–ұв–ұв–ұ', 'в–°в–°в–°в–ұв–ұв–ұв–ұ', 'в–°в–°в–°в–°в–ұв–ұв–ұ', 'в–°в–°в–°в–°в–°в–ұв–ұ', 'в–°в–°в–°в–°в–°в–°в–ұ', 'в–°в–°в–°в–°в–°в–°в–°'], interval: 120 },
  hamburger: { frames: ['вҳұ', 'вҳІ', 'вҳҙ'], interval: 150 },
}

export interface SpinnerProps {
  variant?: SpinnerVariant
  title?: string
  align?: 'left' | 'right'
  speed?: number
}

export function Spinner({ variant = 'dot', title, align = 'left', speed }: SpinnerProps) {
  const { tokens } = useTuiTheme()
  const spinner = SPINNERS[variant]
  const { frames } = spinner
  const interval = speed ?? spinner.interval
  const [frameIndex, setFrameIndex] = useState(0)

  useEffect(() => {
    const id = setInterval(() => {
      setFrameIndex(prev => (prev + 1) % frames.length)
    }, interval)
    return () => clearInterval(id)
  }, [frames.length, interval])

  const frame = frames[frameIndex] ?? frames[0]!

  return (
    <box style={{ flexDirection: 'row', gap: 1 }}>
      {align === 'left' && <text content={frame} fg={tokens.accent} />}
      {title && <text content={title} fg={tokens.textPrimary} />}
      {align === 'right' && <text content={frame} fg={tokens.accent} />}
    </box>
  )
}
