// @bunli/utils/dist/index.js
// @bun
// src/colors.ts
var colorCodes = {
  black: 30,
  red: 31,
  green: 32,
  yellow: 33,
  blue: 34,
  magenta: 35,
  cyan: 36,
  white: 37,
  gray: 90,
  brightRed: 91,
  brightGreen: 92,
  brightYellow: 93,
  brightBlue: 94,
  brightMagenta: 95,
  brightCyan: 96,
  brightWhite: 97,
  bgRed: 41,
  bgGreen: 42,
  bgYellow: 43,
  bgBlue: 44,
  bgMagenta: 45,
  bgCyan: 46,
  bgWhite: 47,
  bold: 1,
  dim: 2,
  italic: 3,
  underline: 4,
  strikethrough: 9,
  reset: 0
};
function createColorFunction(code) {
  return (text) => {
    if (!process.stdout.isTTY || process.env.NO_COLOR) {
      return text;
    }
    return `\x1B[${code}m${text}\x1B[0m`;
  };
}
function stripAnsi(text) {
  return Bun.stripANSI(text);
}
var colors = {
  black: createColorFunction(colorCodes.black),
  red: createColorFunction(colorCodes.red),
  green: createColorFunction(colorCodes.green),
  yellow: createColorFunction(colorCodes.yellow),
  blue: createColorFunction(colorCodes.blue),
  magenta: createColorFunction(colorCodes.magenta),
  cyan: createColorFunction(colorCodes.cyan),
  white: createColorFunction(colorCodes.white),
  gray: createColorFunction(colorCodes.gray),
  brightRed: createColorFunction(colorCodes.brightRed),
  brightGreen: createColorFunction(colorCodes.brightGreen),
  brightYellow: createColorFunction(colorCodes.brightYellow),
  brightBlue: createColorFunction(colorCodes.brightBlue),
  brightMagenta: createColorFunction(colorCodes.brightMagenta),
  brightCyan: createColorFunction(colorCodes.brightCyan),
  brightWhite: createColorFunction(colorCodes.brightWhite),
  bgRed: createColorFunction(colorCodes.bgRed),
  bgGreen: createColorFunction(colorCodes.bgGreen),
  bgYellow: createColorFunction(colorCodes.bgYellow),
  bgBlue: createColorFunction(colorCodes.bgBlue),
  bgMagenta: createColorFunction(colorCodes.bgMagenta),
  bgCyan: createColorFunction(colorCodes.bgCyan),
  bgWhite: createColorFunction(colorCodes.bgWhite),
  bold: createColorFunction(colorCodes.bold),
  dim: createColorFunction(colorCodes.dim),
  italic: createColorFunction(colorCodes.italic),
  underline: createColorFunction(colorCodes.underline),
  strikethrough: createColorFunction(colorCodes.strikethrough),
  reset: createColorFunction(colorCodes.reset),
  strip: stripAnsi
};
// src/shell-integration.ts
async function readStdinLines(delimiter) {
  if (process.stdin.isTTY)
    return [];
  const chunks = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk);
  }
  const data = Buffer.concat(chunks).toString("utf-8");
  if (data.length === 0)
    return [];
  if (delimiter !== undefined) {
    return data.split(delimiter);
  }
  const normalized = data.replace(/\r\n/g, `
`);
  const lines = normalized.split(`
`);
  if (normalized.endsWith(`
`)) {
    lines.pop();
  }
  return lines;
}
function writeStdout(value) {
  const output = process.stdout.isTTY ? value : stripAnsi2(value);
  process.stdout.write(output + `
`);
}
function writeStdoutLines(values) {
  for (const value of values) {
    writeStdout(value);
  }
}
function stripAnsi2(str) {
  return str.replace(/\x1b\[[0-9;]*m/g, "");
}
var EXIT_CODES = {
  SUCCESS: 0,
  CANCEL: 1,
  TIMEOUT: 124,
  SIGINT: 130
};
// src/log.ts
var LEVEL_STYLES = {
  debug: { label: "DEBUG", color: [143, 161, 181] },
  info: { label: "INFO", color: [106, 196, 255] },
  warn: { label: "WARN", color: [249, 200, 91] },
  error: { label: "ERROR", color: [255, 107, 107] },
  fatal: { label: "FATAL", color: [201, 48, 48] }
};
var RESET = "\x1B[0m";
var BOLD = "\x1B[1m";
var DIM = "\x1B[2m";
function bold(s) {
  return `${BOLD}${s}${RESET}`;
}
function dim(s) {
  return `${DIM}${s}${RESET}`;
}
function colored(s, r, g, b) {
  return `\x1B[38;2;${r};${g};${b}m${s}${RESET}`;
}
function formatTimestamp(format) {
  const now = new Date;
  switch (format) {
    case "iso":
      return now.toISOString();
    case "short":
      return now.toISOString().slice(0, 19);
    case "time":
      return now.toTimeString().slice(0, 8);
  }
}
function formatLog(message, options = {}) {
  const {
    level = "info",
    timestamp = false,
    timestampFormat = "short",
    prefix,
    separator = "=",
    fields
  } = options;
  const style = LEVEL_STYLES[level];
  const parts = [];
  if (timestamp) {
    parts.push(dim(formatTimestamp(timestampFormat)));
  }
  const levelStr = colored(style.label, ...style.color);
  if (level === "fatal") {
    parts.push(bold(levelStr));
  } else {
    parts.push(levelStr);
  }
  if (prefix) {
    parts.push(dim(prefix));
  }
  parts.push(message);
  if (fields) {
    for (const [key, value] of Object.entries(fields)) {
      const formattedValue = typeof value === "string" && value.includes(" ") ? `"${value}"` : String(value);
      parts.push(dim(`${key}${separator}${formattedValue}`));
    }
  }
  return parts.join(" ");
}
function log(message, options) {
  process.stderr.write(formatLog(message, options) + `
`);
}
// ../../node_modules/.bun/@standard-schema+utils@0.3.0/node_modules/@standard-schema/utils/dist/index.js
function getDotPath(issue) {
  if (issue.path?.length) {
    let dotPath = "";
    for (const item of issue.path) {
      const key = typeof item === "object" ? item.key : item;
      if (typeof key === "string" || typeof key === "number") {
        if (dotPath) {
          dotPath += `.${key}`;
        } else {
          dotPath += key;
        }
      } else {
        return null;
      }
    }
    return dotPath;
  }
  return null;
}
var SchemaError = class extends Error {
  issues;
  constructor(issues) {
    super(issues[0].message);
    this.name = "SchemaError";
    this.issues = issues;
  }
};

// ../../node_modules/.bun/better-result@2.7.0/node_modules/better-result/dist/index.mjs
function dual(arity, body) {
  if (arity === 2)
    return (...args) => {
      if (args.length >= 2)
        return body(args[0], args[1]);
      return (self) => body(self, args[0]);
    };
  if (arity === 3)
    return (...args) => {
      if (args.length >= 3)
        return body(args[0], args[1], args[2]);
      return (self) => body(self, args[0], args[1]);
    };
  if (arity === 4)
    return (...args) => {
      if (args.length >= 4)
        return body(args[0], args[1], args[2], args[3]);
      return (self) => body(self, args[0], args[1], args[2]);
    };
  return (...args) => {
    if (args.length >= arity)
      return body(...args);
    return (self) => body(self, ...args);
  };
}
var serializeCause = (cause) => {
  if (cause instanceof Error)
    return {
      name: cause.name,
      message: cause.message,
      stack: cause.stack
    };
  return cause;
};
var isAnyTaggedError = (value) => {
  return value instanceof Error && "_tag" in value && typeof value._tag === "string";
};
var TaggedError = Object.assign((tag) => () => {

  class Base extends Error {
    _tag = tag;
    static is(value) {
      return value instanceof Base;
    }
    constructor(args) {
      const message = args && "message" in args && typeof args.message === "string" ? args.message : undefined;
      const cause = args && "cause" in args ? args.cause : undefined;
      super(message, cause !== undefined ? { cause } : undefined);
      if (args)
        Object.assign(this, args);
      Object.setPrototypeOf(this, new.target.prototype);
      this.name = tag;
      if (cause instanceof Error && cause.stack) {
        const indented = cause.stack.replace(/\n/g, `
  `);
        this.stack = `${this.stack}
Caused by: ${indented}`;
      }
    }
    toJSON() {
      return {
        ...this,
        _tag: this._tag,
        name: this.name,
        message: this.message,
        cause: serializeCause(this.cause),
        stack: this.stack
      };
    }
  }
  return Base;
}, { is: isAnyTaggedError });
var matchError = dual(2, (err$1, handlers) => {
  const handler = handlers[err$1._tag];
  return handler(err$1);
});
var matchErrorPartial = dual(3, (err$1, handlers, fallback) => {
  const handler = handlers[err$1._tag];
  if (typeof handler === "function")
    return handler(err$1);
  return fallback(err$1);
});
var UnhandledException = class extends TaggedError("UnhandledException")() {
  constructor(args) {
    const message = args.cause instanceof Error ? `Unhandled exception: ${args.cause.message}` : `Unhandled exception: ${String(args.cause)}`;
    super({
      message,
      cause: args.cause
    });
  }
};
var Panic = class extends TaggedError("Panic")() {
};
var ResultDeserializationError = class extends TaggedError("ResultDeserializationError")() {
  constructor(args) {
    super({
      message: `Failed to deserialize value as Result: expected { status: "ok", value } or { status: "error", error }`,
      value: args.value
    });
  }
};
var panic = (message, cause) => {
  throw new Panic({
    message,
    cause
  });
};
var tryOrPanic = (fn, message) => {
  try {
    return fn();
  } catch (cause) {
    throw panic(message, cause);
  }
};
var tryOrPanicAsync = async (fn, message) => {
  try {
    return await fn();
  } catch (cause) {
    throw panic(message, cause);
  }
};
var Ok = class Ok2 {
  status = "ok";
  constructor(value) {
    this.value = value;
  }
  isOk() {
    return true;
  }
  isErr() {
    return false;
  }
  map(fn) {
    return tryOrPanic(() => new Ok2(fn(this.value)), "map callback threw");
  }
  mapError(_fn) {
    return this;
  }
  andThen(fn) {
    return tryOrPanic(() => fn(this.value), "andThen callback threw");
  }
  andThenAsync(fn) {
    return tryOrPanicAsync(() => fn(this.value), "andThenAsync callback threw");
  }
  match(handlers) {
    return tryOrPanic(() => handlers.ok(this.value), "match ok handler threw");
  }
  unwrap(_message) {
    return this.value;
  }
  unwrapOr(_fallback) {
    return this.value;
  }
  tap(fn) {
    return tryOrPanic(() => {
      fn(this.value);
      return this;
    }, "tap callback threw");
  }
  tapAsync(fn) {
    return tryOrPanicAsync(async () => {
      await fn(this.value);
      return this;
    }, "tapAsync callback threw");
  }
  *[Symbol.iterator]() {
    return this.value;
  }
};
var Err = class Err2 {
  status = "error";
  constructor(error) {
    this.error = error;
  }
  isOk() {
    return false;
  }
  isErr() {
    return true;
  }
  map(_fn) {
    return this;
  }
  mapError(fn) {
    return tryOrPanic(() => new Err2(fn(this.error)), "mapError callback threw");
  }
  andThen(_fn) {
    return this;
  }
  andThenAsync(_fn) {
    return Promise.resolve(this);
  }
  match(handlers) {
    return tryOrPanic(() => handlers.err(this.error), "match err handler threw");
  }
  unwrap(message) {
    return panic(message ?? `Unwrap called on Err: ${String(this.error)}`, this.error);
  }
  unwrapOr(fallback) {
    return fallback;
  }
  tap(_fn) {
    return this;
  }
  tapAsync(_fn) {
    return Promise.resolve(this);
  }
  *[Symbol.iterator]() {
    yield this;
    return panic("Unreachable: Err yielded in Result.gen but generator continued", this.error);
  }
};
function ok(value) {
  return new Ok(value);
}
var isOk = (result) => {
  return result.status === "ok";
};
var err = (error) => new Err(error);
var isError = (result) => {
  return result.status === "error";
};
var tryFn = (options, config) => {
  const execute = () => {
    if (typeof options === "function")
      try {
        return ok(options());
      } catch (cause) {
        return err(new UnhandledException({ cause }));
      }
    try {
      return ok(options.try());
    } catch (originalCause) {
      try {
        return err(options.catch(originalCause));
      } catch (catchHandlerError) {
        throw panic("Result.try catch handler threw", catchHandlerError);
      }
    }
  };
  const times = config?.retry?.times ?? 0;
  let result = execute();
  for (let retry = 0;retry < times && result.status === "error"; retry++)
    result = execute();
  return result;
};
var tryPromise = async (options, config) => {
  const execute = async () => {
    if (typeof options === "function")
      try {
        return ok(await options());
      } catch (cause) {
        return err(new UnhandledException({ cause }));
      }
    try {
      return ok(await options.try());
    } catch (originalCause) {
      try {
        return err(await options.catch(originalCause));
      } catch (catchHandlerError) {
        throw panic("Result.tryPromise catch handler threw", catchHandlerError);
      }
    }
  };
  const retry = config?.retry;
  if (!retry)
    return execute();
  const getDelay = (retryAttempt) => {
    switch (retry.backoff) {
      case "constant":
        return retry.delayMs;
      case "linear":
        return retry.delayMs * (retryAttempt + 1);
      case "exponential":
        return retry.delayMs * 2 ** retryAttempt;
    }
  };
  const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
  let result = await execute();
  const shouldRetryFn = retry.shouldRetry ?? (() => true);
  for (let attempt = 0;attempt < retry.times; attempt++) {
    if (result.status !== "error")
      break;
    const error = result.error;
    if (!tryOrPanic(() => shouldRetryFn(error), "shouldRetry predicate threw"))
      break;
    await sleep(getDelay(attempt));
    result = await execute();
  }
  return result;
};
var map = dual(2, (result, fn) => {
  return result.map(fn);
});
var mapError = dual(2, (result, fn) => {
  return result.mapError(fn);
});
var andThen = dual(2, (result, fn) => {
  return result.andThen(fn);
});
var andThenAsync = dual(2, (result, fn) => {
  return result.andThenAsync(fn);
});
var match = dual(2, (result, handlers) => {
  return result.match(handlers);
});
var tap = dual(2, (result, fn) => {
  return result.tap(fn);
});
var tapAsync = dual(2, (result, fn) => {
  return result.tapAsync(fn);
});
var unwrap = (result, message) => {
  return result.unwrap(message);
};
function assertIsResult(value) {
  if (value !== null && typeof value === "object" && "status" in value && (value.status === "ok" || value.status === "error"))
    return;
  return panic("Result.gen body must return Result.ok() or Result.err(), got: " + (value === null ? "null" : typeof value === "object" ? JSON.stringify(value) : String(value)));
}
var unwrapOr = dual(2, (result, fallback) => {
  return result.unwrapOr(fallback);
});
var gen = (body, thisArg) => {
  const iterator = body.call(thisArg);
  if (Symbol.asyncIterator in iterator)
    return (async () => {
      const asyncIter = iterator;
      let state$1;
      try {
        state$1 = await asyncIter.next();
      } catch (cause) {
        throw panic("generator body threw", cause);
      }
      assertIsResult(state$1.value);
      if (!state$1.done)
        try {
          await asyncIter.return?.(undefined);
        } catch (cause) {
          throw panic("generator cleanup threw", cause);
        }
      return state$1.value;
    })();
  const syncIter = iterator;
  let state;
  try {
    state = syncIter.next();
  } catch (cause) {
    throw panic("generator body threw", cause);
  }
  assertIsResult(state.value);
  if (!state.done)
    try {
      syncIter.return?.(undefined);
    } catch (cause) {
      throw panic("generator cleanup threw", cause);
    }
  return state.value;
};
async function* resultAwait(promise) {
  return yield* await promise;
}
function isSerializedResult(obj) {
  return obj !== null && typeof obj === "object" && "status" in obj && (obj.status === "ok" && ("value" in obj) || obj.status === "error" && ("error" in obj));
}
var serialize = (result) => {
  return result.status === "ok" ? {
    status: "ok",
    value: result.value
  } : {
    status: "error",
    error: result.error
  };
};
var deserialize = (value) => {
  if (isSerializedResult(value))
    return value.status === "ok" ? new Ok(value.value) : new Err(value.error);
  return err(new ResultDeserializationError({ value }));
};
var hydrate = (value) => {
  return deserialize(value);
};
var partition = (results) => {
  const oks = [];
  const errs = [];
  for (const r of results)
    if (r.status === "ok")
      oks.push(r.value);
    else
      errs.push(r.error);
  return [oks, errs];
};
var flatten = (result) => {
  if (result.status === "ok")
    return result.value;
  return result;
};
var Result = {
  ok,
  isOk,
  err,
  isError,
  try: tryFn,
  tryPromise,
  map,
  mapError,
  andThen,
  andThenAsync,
  match,
  tap,
  tapAsync,
  unwrap,
  unwrapOr,
  gen,
  await: resultAwait,
  serialize,
  deserialize,
  hydrate,
  partition,
  flatten
};

// src/validation.ts
async function validate(schema, value) {
  const result = await schema["~standard"].validate(value);
  if (result.issues) {
    return Result.err(new SchemaError(result.issues));
  }
  return Result.ok(result.value);
}
async function validateFields(schemas, values) {
  const results = {};
  const errors = {};
  for (const [rawField, schema] of Object.entries(schemas)) {
    const field = rawField;
    const result = await schema["~standard"].validate(values[rawField]);
    if (result.issues) {
      errors[rawField] = result.issues.map((issue) => issue.message);
    } else {
      results[field] = result.value;
    }
  }
  if (Object.keys(errors).length > 0) {
    return Result.err(errors);
  }
  return Result.ok(results);
}
// src/xdg.ts
import { homedir } from "os";
import { join } from "path";
function getRuntimeEnv() {
  return {
    platform: process.platform,
    env: process.env,
    homedir: homedir()
  };
}
function validateAppName(appName) {
  if (!appName || appName.trim().length === 0) {
    throw new Error("appName must be a non-empty string");
  }
  if (appName.includes("/") || appName.includes("\\")) {
    throw new Error("appName must not contain path separators");
  }
}
function resolveUnixDir(env, xdgEnvVar, fallbackSegments, appName) {
  const xdgValue = env.env[xdgEnvVar];
  const base = xdgValue && xdgValue.trim().length > 0 ? xdgValue : join(env.homedir, ...fallbackSegments);
  return join(base, appName);
}
function resolveWindowsDir(env, envVar, fallbackSegments, appName, ...suffix) {
  const value = env.env[envVar];
  const base = value && value.trim().length > 0 ? value : join(env.homedir, ...fallbackSegments);
  return join(base, appName, ...suffix);
}
function configDir(appName, env) {
  validateAppName(appName);
  const e = env ?? getRuntimeEnv();
  switch (e.platform) {
    case "linux":
    case "darwin":
      return resolveUnixDir(e, "XDG_CONFIG_HOME", [".config"], appName);
    case "win32":
      return resolveWindowsDir(e, "APPDATA", ["AppData", "Roaming"], appName);
    default:
      throw new Error(`Unsupported platform: ${e.platform}`);
  }
}
function dataDir(appName, env) {
  validateAppName(appName);
  const e = env ?? getRuntimeEnv();
  switch (e.platform) {
    case "linux":
    case "darwin":
      return resolveUnixDir(e, "XDG_DATA_HOME", [".local", "share"], appName);
    case "win32":
      return resolveWindowsDir(e, "LOCALAPPDATA", ["AppData", "Local"], appName, "Data");
    default:
      throw new Error(`Unsupported platform: ${e.platform}`);
  }
}
function stateDir(appName, env) {
  validateAppName(appName);
  const e = env ?? getRuntimeEnv();
  switch (e.platform) {
    case "linux":
    case "darwin":
      return resolveUnixDir(e, "XDG_STATE_HOME", [".local", "state"], appName);
    case "win32":
      return resolveWindowsDir(e, "LOCALAPPDATA", ["AppData", "Local"], appName, "State");
    default:
      throw new Error(`Unsupported platform: ${e.platform}`);
  }
}
function cacheDir(appName, env) {
  validateAppName(appName);
  const e = env ?? getRuntimeEnv();
  switch (e.platform) {
    case "linux":
    case "darwin":
      return resolveUnixDir(e, "XDG_CACHE_HOME", [".cache"], appName);
    case "win32":
      return resolveWindowsDir(e, "LOCALAPPDATA", ["AppData", "Local"], appName, "Cache");
    default:
      throw new Error(`Unsupported platform: ${e.platform}`);
  }
}

// src/index.ts
var utils = {
  colors
};
export {
  writeStdoutLines,
  writeStdout,
  validateFields,
  validate,
  utils,
  stripAnsi2 as stripAnsi,
  stateDir,
  readStdinLines,
  log,
  getDotPath,
  formatLog,
  dataDir,
  configDir,
  colors,
  cacheDir,
  SchemaError,
  EXIT_CODES
};
